app/questionnaire/questionnaire_schema.py

```diff
@@ -215,19 +215,40 @@ def get_post_submission(self) -> ImmutableDict:
         schema: ImmutableDict = self.json.get("post_submission", ImmutableDict({}))
         return schema
 
+    def _check_list_exists(self, rules: Union[dict, Sequence], list_name: str) -> bool:
+        # old rules
+        if isinstance(rules, Sequence):
+            return any(
+                rule.get("list") == list_name and "list" in rule for rule in rules
+            )
+
+        # New rules
+        if (
+            isinstance(rules, dict)
+            and rules.get("source") == "list"
+            and rules.get("identifier") == list_name
+        ):
+            return True
```

Comment:
Once we know the source is a `list`, we need to return a bool, otherwise it would error as we are at the final dict and list name didn't match for any number of reasons, such as it is a rule for a diff list name, when this happens, it would currently try to continue iterating and would then fall over.

The test doesn't pick this up because one rule is using the new format, the other rule is using the old format. We should extend it, so you have 4 blocks, 2 that matches the list name and 2 that doesn't. You would use old format for 2 blocks, and new format for the other 2.

```suggestion
        if isinstance(rules, dict) and rules.get("source") == "list":
            return rules["identifier"] == list_name
```